File: spring/member-server/src/main/java/com/example/memberserver/infrastructure/IdProvier/firebase/FirebaseAuthClient.java
package com.example.memberserver.infrastructure.IdProvier.firebase;

import com.example.memberserver.infrastructure.IdProvier.IdpClient;
import com.example.memberserver.member.Member;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthException;
import com.google.firebase.auth.UserRecord;

import java.util.ArrayList;
import java.util.HashMap;

public class FirebaseAuthClient implements IdpClient {

    @Override
    public Long createMember(Member member) {
        UserRecord.CreateRequest createRequest = new UserRecord.CreateRequest();
        createRequest.setUid(member.getId().toString());
        createRequest.setEmail(member.getEmail());
        createRequest.setPassword(member.getPassword());
        try {
            String uid = FirebaseAuth.getInstance().createUser(createRequest).getUid();
            return Long.valueOf(uid);
        } catch (FirebaseAuthException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public Long updateMember(Member member) {
        try {
            FirebaseAuth.getInstance().setCustomUserClaims(member.getId().toString(),
                    new HashMap<>() {{
                        put("permissions", new ArrayList<>(member.getPermissions()));
                    }});
            return member.getId();
        } catch (FirebaseAuthException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void deleteMember(Member member) {
        try {
            FirebaseAuth.getInstance().deleteUser(member.getId().toString());
        } catch (FirebaseAuthException e) {
            e.printStackTrace();
        }
    }
}

File: user-server-firebase-spring-kotlin/src/main/kotlin/com/example/member/FirebaseAuthClient.kt
package com.example.member

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserRecord
import org.springframework.stereotype.Component

@Component
class FirebaseAuthClient: IdpClient {
    override fun createMember(member: Member, password: String): Long? {
        val createRequest = UserRecord.CreateRequest()
        createRequest.setUid(member.id.toString())
        createRequest.setEmail(member.email)
        createRequest.setPassword(password)
        val uid = FirebaseAuth.getInstance().createUser(createRequest).uid
        return uid.toLong()
    }

    override fun updateMember(member: Member): Long? {
        val claims = mapOf("permissions" to member.permissions.map { it.name })
        FirebaseAuth.getInstance().setCustomUserClaims(member.id.toString(), claims)
        return member.id
    }

    override fun deleteMember(member: Member) {
        FirebaseAuth.getInstance().deleteUser(member.id.toString())
    }
}
